The ``lazy_matrix_mul`` module
======================

Importing ``lazy_matrix_mul``
-------------------

    >>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul

Testing ``lazy_matrix_mul``
-------------------

both matrices contain only ints
    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]])
    array([[ 7, 10],
           [15, 22]])

both matrices contain only floats
    >>> lazy_matrix_mul([[1., 2.], [3., 4.]], [[1., 2.], [3., 4.]])
    array([[ 7., 10.],
           [15., 22.]])

both matrices contain ints and floats
    >>> lazy_matrix_mul([[1., 2], [3, 4]], [[1., 2], [3, 4]])
    array([[ 7., 10.],
           [15., 22.]])

m_a is a str
    >>> lazy_matrix_mul("[[1., 2], [3, 4]]", [[1., 2], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: Scalar operands are not allowed, use '*' instead

m_b is a str
    >>> lazy_matrix_mul([[1., 2], [3, 4]], "[[1., 2], [3, 4]]")
    Traceback (most recent call last):
        ...
    TypeError: Scalar operands are not allowed, use '*' instead

m_a is a list
    >>> lazy_matrix_mul([1, 2, 3, 4], [[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: m_a must be a list of lists

m_b is a list
    >>> lazy_matrix_mul([[1, 2], [3, 4]], [1, 2, 3, 4])
    Traceback (most recent call last):
        ...
    TypeError: m_b must be a list of lists

m_a is a empty
    >>> lazy_matrix_mul([[]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    ValueError: shapes (1,0) and (2,2) not aligned: 0 (dim 1) != 2 (dim 0)

m_b is a empty
    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[]])
    Traceback (most recent call last):
        ...
    ValueError: shapes (2,2) and (1,0) not aligned: 2 (dim 1) != 1 (dim 0)

m_a contains non-numbers
    >>> lazy_matrix_mul([["1", 2], [3, 4]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: invalid data type for einsum

m_b contains non-numbers
    >>> lazy_matrix_mul([[1, 2], [3, 4]], [["1", 2], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: invalid data type for einsum

m_a has mismatched rows
    >>> lazy_matrix_mul([[1, 2], [2]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: setting an array element with a sequence.

m_b has mismatched rows
    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [2]])
    Traceback (most recent call last):
        ...
    TypeError: setting an array element with a sequence.

m_a and m_b can't be multiplied
    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1]])
    Traceback (most recent call last):
        ...
    ValueError: shapes (2,2) and (1,1) not aligned: 2 (dim 1) != 1 (dim 0)

missing one argument
    >>> lazy_matrix_mul([[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

missing two arguments
    >>> lazy_matrix_mul()
    Traceback (most recent call last):
        ...
    TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'